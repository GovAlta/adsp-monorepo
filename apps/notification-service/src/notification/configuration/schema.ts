import { Channel, NotificationType } from '../types';

const templateSchema = {
  type: 'object',
  properties: {
    subject: { type: 'string' },
    body: { type: 'string' },
  },
  required: ['subject', 'body'],
};

export const configurationSchema = {
  type: 'object',
  properties: {
    contact: {
      type: 'object',
      properties: {
        contactEmail: { type: 'string' },
        phoneNumber: { type: 'string' },
        supportInstructions: { type: 'string' },
      },
    },
  },
  additionalProperties: {
    type: 'object',
    properties: {
      id: { type: 'string', pattern: '^[a-zA-Z0-9-_ ]{1,50}$' },
      name: { type: 'string' },
      description: { type: ['string', 'null'] },
      publicSubscribe: { type: 'boolean' },
      manageSubscribe: { type: 'boolean' },
      channels: {
        type: 'array',
        items: {
          type: 'string',
          enum: Object.values(Channel),
        },
      },
      subscriberRoles: {
        type: 'array',
        items: { type: 'string' },
      },
      events: {
        type: 'array',
        items: {
          type: 'object',
          properties: {
            namespace: { type: 'string', pattern: '^[a-zA-Z0-9-_ ]{1,50}$' },
            name: { type: 'string', pattern: '^[a-zA-Z0-9-_ ]{1,50}$' },
            templates: {
              type: 'object',
              properties: {
                email: templateSchema,
                sms: templateSchema,
                mail: templateSchema,
                bot: templateSchema,
              },
            },
          },
          required: ['namespace', 'name', 'templates'],
        },
      },
    },
    required: ['id', 'name', 'publicSubscribe', 'subscriberRoles', 'events'],
  },
};

export interface SupportContact {
  contactEmail: string;
  phoneNumber: string;
  supportInstructions: string;
}

export type Configuration = {
  contact?: SupportContact;
} & Record<string, NotificationType>;
