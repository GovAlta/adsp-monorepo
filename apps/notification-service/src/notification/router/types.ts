import { Channel } from '@abgov/adsp-service-sdk';

export const SUBSCRIBER_SEND_VERIFY_CODE = 'send-code';
export interface SendVerifyCodeRequest {
  operation: typeof SUBSCRIBER_SEND_VERIFY_CODE;
  channel: Channel;
  address: string;
  reason?: string;
}

export const SUBSCRIBER_CHECK_CODE = 'check-code';
export interface CheckVerifyCodeRequest {
  operation: typeof SUBSCRIBER_CHECK_CODE;
  channel: Channel;
  address: string;
  code: string;
}

export const SUBSCRIBER_VERIFY_CHANNEL = 'verify-channel';
export interface VerifyChannelRequest {
  operation: typeof SUBSCRIBER_VERIFY_CHANNEL;
  channel: Channel;
  address: string;
  code: string;
}

export type SubscriberOperationRequests = SendVerifyCodeRequest | CheckVerifyCodeRequest | VerifyChannelRequest;
