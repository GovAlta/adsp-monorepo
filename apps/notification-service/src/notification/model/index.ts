export * from './space';
export * from './type';
export * from './subscriber';
export * from './subscription';
