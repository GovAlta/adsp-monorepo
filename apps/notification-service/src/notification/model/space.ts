import { AssertRole } from '@abgov/adsp-service-sdk';
import type { User } from '@abgov/adsp-service-sdk';
import { UnauthorizedError, Update } from '@core-services/core-common';
import type { NotificationSpaceRepository } from '../repository';
import type { NotificationSpace } from '../types';
import { ServiceUserRoles } from '../types';

export class NotificationSpaceEntity implements NotificationSpace {
  id: string;
  name: string;
  spaceAdminRole: string;
  subscriberAdminRole: string;

  @AssertRole('create notification space', ServiceUserRoles.Admin)
  static create(user: User, repository: NotificationSpaceRepository, space: NotificationSpace) {
    const entity = new NotificationSpaceEntity(repository, space);
    return repository.save(entity);
  }

  constructor(private repository: NotificationSpaceRepository, space: NotificationSpace) {
    this.id = space.id;
    this.name = space.name;
    this.spaceAdminRole = space.spaceAdminRole;
    this.subscriberAdminRole = space.subscriberAdminRole;
  }

  update(user: User, update: Update<NotificationSpace>) {
    if (!this.canUpdate(user)) {
      throw new UnauthorizedError('User not authorized to updated space.');
    }

    if (update.name) {
      this.name = update.name;
    }

    if (update.spaceAdminRole) {
      this.spaceAdminRole = update.spaceAdminRole;
    }

    if (update.subscriberAdminRole) {
      this.subscriberAdminRole = update.subscriberAdminRole;
    }

    return this.repository.save(this);
  }

  canAccess(user: User) {
    return user && (user.roles.includes(ServiceUserRoles.Admin) || user.roles.includes(this.spaceAdminRole));
  }

  canUpdate(user: User) {
    return user && (user.roles.includes(ServiceUserRoles.Admin) || user.roles.includes(this.spaceAdminRole));
  }
}
