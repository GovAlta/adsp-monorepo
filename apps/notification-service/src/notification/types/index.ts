export * from './space';
export * from './user';
export * from './type';
export * from './subscription';
export * from './subscriber';
export * from './notification';
export * from './provider';
export * from './template';
export * from './message';
export * from './channel';
