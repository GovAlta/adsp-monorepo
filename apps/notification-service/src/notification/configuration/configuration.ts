import type { AdspId } from '@abgov/adsp-service-sdk';
import type { DomainEvent } from '@core-services/core-common';
import { NotificationType } from '../types';
import { NotificationTypeEntity } from '../model';

export class NotificationConfiguration {
  private types: Record<string, NotificationTypeEntity>;
  private eventTypes: Record<string, NotificationTypeEntity[]>;
  constructor(types: Record<string, NotificationType>, tenantId?: AdspId) {
    types = types || {};
    this.types = Object.keys(types)
      .filter((type) => type !== 'contact')
      .reduce((entities, key) => {
        entities[key] = new NotificationTypeEntity(types[key], tenantId);
        return entities;
      }, {});

    this.eventTypes = Object.keys(this.types).reduce((eventEntities, key) => {
      const type = this.types[key];
      type.events.forEach(({ namespace, name }) => {
        const eventKey = this.getEventKey(namespace, name);
        eventEntities[eventKey] = [...(eventEntities[eventKey] || []), type];
      });
      return eventEntities;
    }, {} as Record<string, NotificationTypeEntity[]>);
  }

  private getEventKey(namespace: string, name: string) {
    return `${namespace}:${name}`;
  }

  getNotificationTypes(): NotificationTypeEntity[] {
    return Object.keys(this.types).map((k) => this.types[k]);
  }

  getNotificationType(type: string): NotificationTypeEntity {
    return this.types[type];
  }

  getEventNotificationTypes({ namespace, name }: DomainEvent): NotificationTypeEntity[] {
    const key = this.getEventKey(namespace, name);
    return this.eventTypes[key] || [];
  }
}
