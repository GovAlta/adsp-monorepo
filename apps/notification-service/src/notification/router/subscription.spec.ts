import { adspId } from '@abgov/adsp-service-sdk';
import { NotFoundError, UnauthorizedError } from '@core-services/core-common';
import { Request, Response } from 'express';
import { Logger } from 'winston';
import { NotificationTypeEntity, SubscriberEntity, SubscriptionEntity } from '../model';
import { NotificationConfiguration } from '../configuration';
import {
  createTypeSubscription,
  getTypeSubscriptions,
  addTypeSubscription,
  createSubscriptionRouter,
  deleteTypeSubscription,
  getNotificationType,
  getNotificationTypes,
  getSubscriber,
  getSubscribers,
  getTypeSubscription,
} from './subscription';
import { NotificationType, ServiceUserRoles } from '../types';
import { deleteSubscriber, updateSubscriber } from '.';

describe('subscription router', () => {
  const tenantId = adspId`urn:ads:platform:tenant-service:v2:/tenants/test`;

  const loggerMock = {
    debug: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
  } as unknown as Logger;

  const repositoryMock = {
    getSubscriber: jest.fn(),
    getSubscriptions: jest.fn(),
    getSubscription: jest.fn(),
    findSubscribers: jest.fn(),
    deleteSubscriber: jest.fn(),
    saveSubscription: jest.fn(),
    deleteSubscriptions: jest.fn(),
    saveSubscriber: jest.fn((entity) => Promise.resolve(entity)),
  };

  const notificationType: NotificationType = {
    id: 'test',
    name: 'Test',
    description: 'testing 1 2 3',
    events: [],
    subscriberRoles: ['test-subscriber'],
  };

  beforeEach(() => {
    repositoryMock.findSubscribers.mockReset();
    repositoryMock.getSubscriber.mockReset();
    repositoryMock.getSubscriptions.mockReset();
    repositoryMock.saveSubscriber.mockClear();
    repositoryMock.saveSubscription.mockReset();
    repositoryMock.deleteSubscriptions.mockReset();
    repositoryMock.deleteSubscriber.mockReset();
  });

  describe('createSubscriptionRouter', () => {
    it('can create router', () => {
      const router = createSubscriptionRouter({ logger: loggerMock, subscriptionRepository: repositoryMock });
      expect(router).toBeTruthy();
    });
  });

  describe('getNotificationTypes', () => {
    it('can get types', async () => {
      const req = {
        getConfiguration: jest.fn(),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      req.getConfiguration.mockResolvedValueOnce([new NotificationConfiguration({ test: notificationType }, tenantId)]);
      await getNotificationTypes(req as unknown as Request, res as unknown as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.arrayContaining([expect.objectContaining(notificationType)]));
    });
  });

  describe('getNotificationType', () => {
    it('can get type', async () => {
      const req = {
        params: { type: 'test' },
        getConfiguration: jest.fn(),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      req.getConfiguration.mockResolvedValueOnce([new NotificationConfiguration({ test: notificationType }, tenantId)]);
      await getNotificationType(req as unknown as Request, res as unknown as Response, next);
      expect(req['notificationType']).toMatchObject(notificationType);
      expect(res.send).not.toHaveBeenCalled();
      expect(next).toHaveBeenCalledWith();
    });

    it('can call next with not found', async () => {
      const req = {
        params: { type: 'not-there' },
        getConfiguration: jest.fn(),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      req.getConfiguration.mockResolvedValueOnce([new NotificationConfiguration({ test: notificationType }, tenantId)]);
      await getNotificationType(req as unknown as Request, res as unknown as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(NotFoundError));
    });
  });

  describe('getTypeSubscriptions', () => {
    it('can create handler', () => {
      const handler = getTypeSubscriptions(repositoryMock);
      expect(handler).toBeTruthy();
    });

    it('can get subscriptions', async () => {
      const req = {
        query: {},
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      const result = { results: [], page: {} };
      repositoryMock.getSubscriptions.mockResolvedValueOnce(result);

      const handler = getTypeSubscriptions(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ page: result.page }));
    });

    it('can get handle query params', async () => {
      const req = {
        query: { top: '11', after: '123' },
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      const result = { results: [], page: {} };
      repositoryMock.getSubscriptions.mockResolvedValueOnce(result);

      const handler = getTypeSubscriptions(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(repositoryMock.getSubscriptions).toHaveBeenCalledWith(req.notificationType, 11, '123');
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ page: result.page }));
    });
  });

  describe('createTypeSubscription', () => {
    it('can create handler', () => {
      const handler = createTypeSubscription(repositoryMock);
      expect(handler).toBeTruthy();
    });

    it('can create subscription', async () => {
      const req = {
        user: {
          tenantId,
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        query: {},
        body: { addressAs: 'tester' },
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      const subscriber = new SubscriberEntity(repositoryMock, {
        id: 'subscriber',
        tenantId,
        addressAs: 'tester',
        channels: [],
      });
      const subscription = new SubscriptionEntity(
        repositoryMock,
        { tenantId, typeId: 'test', subscriberId: 'subscriber', criteria: {} },
        subscriber
      );
      repositoryMock.saveSubscriber.mockResolvedValueOnce(subscriber);
      repositoryMock.saveSubscription.mockResolvedValueOnce(subscription);

      const handler = createTypeSubscription(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(repositoryMock.saveSubscriber).toHaveBeenCalledWith(expect.objectContaining({ addressAs: 'tester' }));
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ typeId: 'test' }));
    });

    it('can create user subscription', async () => {
      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        query: { userSub: 'true' },
        body: {},
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      const subscriber = new SubscriberEntity(repositoryMock, {
        id: 'subscriber',
        tenantId,
        addressAs: 'tester',
        channels: [],
      });
      const subscription = new SubscriptionEntity(
        repositoryMock,
        { tenantId, typeId: 'test', subscriberId: 'subscriber', criteria: {} },
        subscriber
      );
      repositoryMock.saveSubscriber.mockResolvedValueOnce(subscriber);
      repositoryMock.saveSubscription.mockResolvedValueOnce(subscription);

      const handler = createTypeSubscription(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(repositoryMock.saveSubscriber).toHaveBeenCalledWith(
        expect.objectContaining({
          addressAs: 'Tester',
          userId: 'tester',
          channels: expect.arrayContaining([expect.objectContaining({ channel: 'email', address: 'tester@test.co' })]),
        })
      );
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ typeId: 'test' }));
    });

    it('can create subscription for existing subscriber', async () => {
      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        query: { userSub: 'true' },
        body: {},
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      const subscriber = new SubscriberEntity(repositoryMock, {
        id: 'subscriber',
        tenantId,
        addressAs: 'tester',
        channels: [],
      });
      const subscription = new SubscriptionEntity(
        repositoryMock,
        { tenantId, typeId: 'test', subscriberId: 'subscriber', criteria: {} },
        subscriber
      );
      repositoryMock.getSubscriber.mockResolvedValueOnce(subscriber);
      repositoryMock.saveSubscription.mockResolvedValueOnce(subscription);

      const handler = createTypeSubscription(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(repositoryMock.getSubscriber).toHaveBeenCalledWith(tenantId, 'tester', true);
      expect(repositoryMock.saveSubscriber).not.toBeCalled();
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ typeId: 'test' }));
    });
  });

  describe('addTypeSubscription', () => {
    it('can create handler', () => {
      const handler = addTypeSubscription(repositoryMock);
      expect(handler).toBeTruthy();
    });

    it('can add subscription', async () => {
      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        query: {},
        params: { subscriber: 'subscriber' },
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      const subscriber = new SubscriberEntity(repositoryMock, {
        id: 'subscriber',
        tenantId,
        addressAs: 'tester',
        channels: [],
      });
      const subscription = new SubscriptionEntity(
        repositoryMock,
        { tenantId, typeId: 'test', subscriberId: 'subscriber', criteria: {} },
        subscriber
      );
      repositoryMock.getSubscriber.mockResolvedValueOnce(subscriber);
      repositoryMock.saveSubscription.mockResolvedValueOnce(subscription);

      const handler = addTypeSubscription(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(repositoryMock.getSubscriber).toHaveBeenCalledWith(tenantId, 'subscriber', false);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ typeId: 'test' }));
    });

    it('can call next with not found', async () => {
      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        query: {},
        params: { subscriber: 'subscriber' },
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      repositoryMock.getSubscriber.mockResolvedValueOnce(null);

      const handler = addTypeSubscription(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(next).toBeCalledWith(expect.any(NotFoundError));
    });
  });

  describe('getTypeSubscription', () => {
    it('can create handler', () => {
      const handler = getTypeSubscription(repositoryMock);
      expect(handler).toBeTruthy();
    });

    it('can get subscription', async () => {
      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        query: {},
        params: { subscriber: 'subscriber' },
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      const subscriber = new SubscriberEntity(repositoryMock, {
        id: 'subscriber',
        tenantId,
        addressAs: 'tester',
        channels: [],
      });
      const subscription = new SubscriptionEntity(
        repositoryMock,
        { tenantId, typeId: 'test', subscriberId: 'subscriber', criteria: {} },
        subscriber
      );

      repositoryMock.getSubscription.mockResolvedValueOnce(subscription);

      const handler = getTypeSubscription(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(repositoryMock.getSubscription).toHaveBeenCalledWith(req.notificationType, 'subscriber');
      expect(res.send).toHaveBeenCalledWith(
        expect.objectContaining({ typeId: 'test', subscriber: expect.objectContaining({ id: 'subscriber' }) })
      );
    });
  });

  describe('deleteTypeSubscription', () => {
    it('can create handler', () => {
      const handler = deleteTypeSubscription(repositoryMock);
      expect(handler).toBeTruthy();
    });

    it('can delete subscription', async () => {
      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        query: {},
        params: { subscriber: 'subscriber' },
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      const subscriber = new SubscriberEntity(repositoryMock, {
        id: 'subscriber',
        tenantId,
        addressAs: 'tester',
        channels: [],
      });

      repositoryMock.getSubscriber.mockResolvedValueOnce(subscriber);
      repositoryMock.deleteSubscriptions.mockResolvedValueOnce(true);

      const handler = deleteTypeSubscription(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(repositoryMock.deleteSubscriptions).toHaveBeenCalledWith(tenantId, notificationType.id, 'subscriber');
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ deleted: true }));
    });
  });

  describe('getSubscribers', () => {
    it('can create handler', () => {
      const handler = getSubscribers(repositoryMock);
      expect(handler).toBeTruthy();
    });

    it('can get subscribers', async () => {
      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        query: {},
        params: { subscriber: 'subscriber' },
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      const result = { results: [], page: {} };
      repositoryMock.findSubscribers.mockResolvedValueOnce(result);

      const handler = getSubscribers(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ page: result.page }));
    });

    it('can get subscribers with query params', async () => {
      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        query: { top: '11', after: '123' },
        params: { subscriber: 'subscriber' },
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      const result = { results: [], page: {} };
      repositoryMock.findSubscribers.mockResolvedValueOnce(result);

      const handler = getSubscribers(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(repositoryMock.findSubscribers).toHaveBeenCalledWith(
        11,
        '123',
        expect.objectContaining({ tenantIdEquals: tenantId })
      );
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ page: result.page }));
    });

    it('can call next with error for non-admin', async () => {
      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [],
        },
        query: {},
        params: { subscriber: 'subscriber' },
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      const handler = getSubscribers(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(UnauthorizedError));
    });
  });

  describe('getSubscriber', () => {
    it('can create handler', () => {
      const handler = getSubscriber(repositoryMock);
      expect(handler).toBeTruthy();
    });

    it('can get subscriber', async () => {
      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        query: {},
        params: { subscriber: 'subscriber' },
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      const subscriber = new SubscriberEntity(repositoryMock, {
        id: 'subscriber',
        tenantId,
        addressAs: 'tester',
        channels: [],
      });

      repositoryMock.getSubscriber.mockResolvedValueOnce(subscriber);

      const handler = getSubscriber(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(repositoryMock.getSubscriber).toHaveBeenCalledWith(tenantId, 'subscriber');
      expect(req['subscriber']).toBe(subscriber);
    });

    it('can call next with not found.', async () => {
      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        query: {},
        params: { subscriber: 'subscriber' },
        notificationType: new NotificationTypeEntity(notificationType, tenantId),
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      repositoryMock.getSubscriber.mockResolvedValueOnce(null);

      const handler = getSubscriber(repositoryMock);
      await handler(req as unknown as Request, res as unknown as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(NotFoundError));
    });
  });

  describe('deleteSubscriber', () => {
    it('can delete subscriber', async () => {
      const subscriber = new SubscriberEntity(repositoryMock, {
        id: 'subscriber',
        tenantId,
        addressAs: 'tester',
        channels: [],
      });

      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        query: {},
        subscriber,
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      repositoryMock.deleteSubscriber.mockResolvedValueOnce(true);

      await deleteSubscriber(req as unknown as Request, res as unknown as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ deleted: true }));
    });

    it('can call next with unauthorized', async () => {
      const subscriber = new SubscriberEntity(repositoryMock, {
        id: 'subscriber',
        tenantId,
        addressAs: 'tester',
        channels: [],
      });

      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [],
        },
        query: {},
        subscriber,
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      await deleteSubscriber(req as unknown as Request, res as unknown as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(UnauthorizedError));
      expect(res.send).not.toHaveBeenCalled();
    });
  });

  describe('updateSubscriber', () => {
    it('can delete subscriber', async () => {
      const subscriber = new SubscriberEntity(repositoryMock, {
        id: 'subscriber',
        tenantId,
        addressAs: 'tester',
        channels: [],
      });

      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [ServiceUserRoles.SubscriptionAdmin],
        },
        body: { addressAs: 'Best Tester' },
        subscriber,
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      repositoryMock.saveSubscriber.mockResolvedValueOnce(subscriber);

      await updateSubscriber(req as unknown as Request, res as unknown as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining(req.body));
    });

    it('can call next with unauthorized', async () => {
      const subscriber = new SubscriberEntity(repositoryMock, {
        id: 'subscriber',
        tenantId,
        addressAs: 'tester',
        channels: [],
      });

      const req = {
        user: {
          id: 'tester',
          tenantId,
          name: 'Tester',
          email: 'tester@test.co',
          roles: [],
        },
        body: {},
        subscriber,
      };
      const res = { send: jest.fn() };
      const next = jest.fn();

      await updateSubscriber(req as unknown as Request, res as unknown as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(UnauthorizedError));
      expect(res.send).not.toHaveBeenCalled();
    });
  });
});
