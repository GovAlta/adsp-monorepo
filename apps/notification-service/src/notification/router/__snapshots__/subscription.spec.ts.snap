// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscription router addOrUpdateTypeSubscription can add subscription 1`] = `
Object {
  "criteria": Array [
    Object {},
  ],
  "subscriber": Object {
    "addressAs": "tester",
    "channels": Array [],
    "id": "subscriber",
    "tenantId": "urn:ads:platform:tenant-service:v2:/tenants/test",
    "urn": "urn:ads:platform:notification-service:v1:/subscribers/subscriber",
    "userId": undefined,
  },
  "subscriberId": "subscriber",
  "type": undefined,
  "typeId": "test",
}
`;

exports[`subscription router addOrUpdateTypeSubscription can update subscription criteria 1`] = `
Object {
  "criteria": Array [
    Object {
      "context": Object {},
      "correlationId": "123",
    },
  ],
  "subscriber": Object {
    "addressAs": "tester",
    "channels": Array [],
    "id": "subscriber",
    "tenantId": "urn:ads:platform:tenant-service:v2:/tenants/test",
    "urn": "urn:ads:platform:notification-service:v1:/subscribers/subscriber",
    "userId": undefined,
  },
  "subscriberId": "subscriber",
  "type": undefined,
  "typeId": "test",
}
`;

exports[`subscription router createSubscriber can create subscriber 1`] = `
Object {
  "addressAs": "tester@test.com",
  "channels": Array [],
  "id": "subscriber",
  "tenantId": "urn:ads:platform:tenant-service:v2:/tenants/test",
  "urn": "urn:ads:platform:notification-service:v1:/subscribers/subscriber",
  "userId": undefined,
}
`;

exports[`subscription router createTypeSubscription can create subscription 1`] = `
Object {
  "criteria": Array [
    Object {},
  ],
  "subscriber": Object {
    "addressAs": "tester",
    "channels": Array [],
    "id": "subscriber",
    "tenantId": "urn:ads:platform:tenant-service:v2:/tenants/test",
    "urn": "urn:ads:platform:notification-service:v1:/subscribers/subscriber",
    "userId": undefined,
  },
  "subscriberId": "subscriber",
  "type": undefined,
  "typeId": "test",
}
`;

exports[`subscription router deleteSubscriber can delete subscriber 1`] = `
Object {
  "deleted": true,
}
`;

exports[`subscription router deleteTypeSubscription can delete subscription 1`] = `
Object {
  "deleted": true,
}
`;

exports[`subscription router getNotificationTypes can get types 1`] = `
Array [
  Object {
    "channels": Array [],
    "description": "testing 1 2 3",
    "events": Array [],
    "id": "test",
    "manageSubscribe": false,
    "name": "Test",
    "publicSubscribe": false,
    "subscriberRoles": Array [
      "test-subscriber",
    ],
  },
]
`;

exports[`subscription router getSubscriberDetails can get subscriber subscriptions 1`] = `
Object {
  "addressAs": "tester",
  "channels": Array [],
  "id": "subscriber",
  "subscriptions": Array [
    Object {
      "criteria": Array [
        Object {},
      ],
      "subscriberId": "subscriber",
      "type": Object {
        "channels": Array [],
        "description": "testing 1 2 3",
        "id": "test",
        "manageSubscribe": false,
        "name": "Test",
        "publicSubscribe": false,
      },
      "typeId": "test",
    },
  ],
  "tenantId": "urn:ads:platform:tenant-service:v2:/tenants/test",
  "urn": "urn:ads:platform:notification-service:v1:/subscribers/subscriber",
  "userId": undefined,
}
`;

exports[`subscription router getSubscriberSubscriptions can get subscriptions 1`] = `
Object {
  "page": Object {},
  "results": Array [],
}
`;

exports[`subscription router getSubscribers can get subscribers 1`] = `
Object {
  "page": Object {},
  "results": Array [],
}
`;

exports[`subscription router getSubscriptionChannels can get channels 1`] = `
Array [
  Object {
    "address": "tester@test.co",
    "channel": "email",
    "verified": false,
  },
]
`;

exports[`subscription router getTypeSubscription can get subscription 1`] = `
Object {
  "criteria": Array [
    Object {},
  ],
  "subscriber": Object {
    "addressAs": "tester",
    "channels": Array [],
    "id": "subscriber",
    "tenantId": "urn:ads:platform:tenant-service:v2:/tenants/test",
    "urn": "urn:ads:platform:notification-service:v1:/subscribers/subscriber",
    "userId": undefined,
  },
  "subscriberId": "subscriber",
  "type": undefined,
  "typeId": "test",
}
`;

exports[`subscription router getTypeSubscriptions can get subscriptions 1`] = `
Object {
  "page": Object {},
  "results": Array [],
}
`;

exports[`subscription router subscriberOperations can check code 1`] = `
Object {
  "verified": true,
}
`;

exports[`subscription router subscriberOperations can send code 1`] = `
Object {
  "sent": true,
}
`;

exports[`subscription router subscriberOperations can verify channel 1`] = `
Object {
  "verified": true,
}
`;

exports[`subscription router updateSubscriber can delete subscriber 1`] = `
Object {
  "addressAs": "tester",
  "channels": Array [],
  "id": "subscriber",
  "tenantId": "urn:ads:platform:tenant-service:v2:/tenants/test",
  "urn": "urn:ads:platform:notification-service:v1:/subscribers/subscriber",
  "userId": undefined,
}
`;
